src/backend/distributed/clock/README.md

```diff
@@ -0,0 +1,244 @@
+### Definitions
+**Total ordering** - Every pair of change events can be placed in some order.
+**Causal ordering** - Only events that are causally related (an event A caused an event B) can be ordered i.e., it's only a partial order - sometimes events happen independently with no possible causal relationship, such events are treated to concurrent.
+**Sequential consistency** - All writes must be seen in the same order by all processes.
+**Causal consistency** - Causally related writes must be seen in the same order.
+
+Transactions on a single node system naturally provide a total and sequential ordering guarantees for client read and write operations as all operations are routed to the same node, but there are challenges for a multi node distributed system, such as, Citus.
+
+One possible way to totally order all the changes in the system is to timestamp all the events with a global physical clock or a centralized logical clock. Thus, observing the events in the increasing order of the timestamp will give the total ordering of events. For both the performance and cost reasons such solutions are impractical. In the absence of total ordering, a little weaker ordering is the **causal order**.
+
+Causal order is defined as a model that preserves a partial order of events in a distributed system. If an event
+
+1.  A causes another event B, every other process in the system observes the event A before observing event B.
+2.  Causal order is transitive: if A causes B, and B causes C, then A causes C.
+3.  Non causally ordered events are treated as concurrent.
+
+Causal consistency is a weak form of consistency that preserves the order of causally related operations. The causal consistency model can be refined into four session guarantees.
+
+1.  Read Your Writes: If a process performs a write, the same process later observes the result of its write.
+6.  Monotonic Reads: The set of writes observed (read) by a process is guaranteed to be monotonically increasing.
+7.  Writes Follow Reads: If some process performs a read followed by a write, and another process observes the result of the write, then it can also observe the read.
+8.  Monotonic Writes: If some process performs a write, followed sometime later by another write, other processes will observe them in the same order.
+
+### Hybrid Logical Clock (HLC)
+HLC provides a way to get the causality relationship like logical clocks. It can also be used for backup/recovery too as the logical clock value is maintained close to the wall clock time. HLC consists of
+ LC - Logical clock
+ C - Counter
+
+Clock components - Unsigned 64 bit <LC, C>
+Epoch Milliseconds ( LC ) | Logical counter ( C )|
+|--|--|
+| 42 bits | 22 bits |
+
+2^42 milliseconds - 4398046511104 milliseconds, which is ~139 years.
+
+2^22 ticks - maximum of four million operations per millisecond.
```

Comment:
Yes, they do, and we have unit tests (in this PR); having said that, we have the luxury of expanding these limits if we need to. BTW, we are limiting the counter not to tick beyond the 22 bits but there is no such logic for logical clock (I don't see how it can cross that limit unless something really went bad), it gets checked only while using UDFs i.e. if any logical clock input with the value > 42 bit max, we will raise an exception.

![image](https://user-images.githubusercontent.com/44680808/198687059-5b33a6f8-a882-4aeb-a0f5-88e68bcfd052.png)